### PR TITLE
Bow side loss

### PR DESCRIPTION
**Pull request by vitaka**
Bag-of-words side loss for AEVNMT.

Enable SL BoW side loss with --bow_loss true
Enable TL BoW side loss with --bow_loss_tl true

For each language (SL and TL) a FFNN followed by a softmax activation function generates a probability distribution of the words in the vocabulary. Loss is simply a cross-entropy loss using as a reference the words present in the sentence. The padding token us ignored.

Numerical stability issues have been detected when training with num_annealing steps < 30,000. The scale of the approximate posterior approaches too much to zero.

### CODE REVIEW

File: aevnmt/aevnmt_helper.py [hunk 1]
```diff
@@ -52,7 +52,7 @@ def create_model(hparams, vocab_src, vocab_tgt):
                    language_model=rnnlm,
                    pad_idx=vocab_tgt[PAD_TOKEN],
                    dropout=hparams.dropout,
-                   tied_embeddings=hparams.tied_embeddings)
+                   tied_embeddings=hparams.tied_embeddings,bow=hparams.bow_loss,bow_tl=hparams.bow_loss_tl)
```
Comment (Roxot) [inline aevnmt/aevnmt_helper.py]:
```suggestion
                   tied_embeddings=hparams.tied_embeddings,
                   bow=hparams.bow_loss,
                   bow_tl=hparams.bow_loss_tl)
```

File: aevnmt/aevnmt_helper.py [hunk 2]
```diff
@@ -75,7 +75,7 @@ def train_step(model, x_in, x_out, seq_mask_x, seq_len_x, noisy_x_in, y_in, y_ou
     loss = model.loss(tm_logits, lm_logits, y_out, x_out, qz,
                       free_nats=hparams.KL_free_nats,
                       KL_weight=KL_weight,
-                      reduction="mean")
+                      reduction="mean",bow_logits=bow_logits,bow_logits_tl=bow_logits_tl)
```
Comment (Roxot) [inline aevnmt/aevnmt_helper.py]:
```suggestion
                      reduction="mean",
                      bow_logits=bow_logits,
                      bow_logits_tl=bow_logits_tl)
```

File: aevnmt/models/aevnmt.py [hunk 3]
```diff
@@ -78,7 +90,10 @@ def inference_parameters(self):
     def generative_parameters(self):
         # TODO: separate the generative model into a GenerativeModel module
         #  within that module, have two modules, namely, LanguageModel and TranslationModel
-        return chain(self.lm_parameters(), self.tm_parameters())
+        return chain(self.lm_parameters(), self.tm_parameters(), self.bow_parameters())
+
+    def bow_parameters(self):
+        return chain( iter(()) if self.bow_output_layer is None else self.bow_output_layer.parameters()   , iter(()) if self.bow_output_layer_tl is None else self.bow_output_layer_tl.parameters()  )
```
Comment (Roxot) [inline aevnmt/models/aevnmt.py]: Please write this in regular if-statements as this is quite hard to follow.

**Review by Roxot — CHANGES_REQUESTED** [review]
Nice, besides some minor style comments this can be merged.

File: aevnmt/models/aevnmt.py (hunk 3 above)
Comment (vitaka) [reply aevnmt/models/aevnmt.py]: Done, added new commit with regular ifs.

**Comment by vitaka** [comment]
I think everything is ready for merging now.
Once this pull request is accepted, I will fix the conflicts in the other remaining pull request, since the conflicting lines will change after merging this branch.
